### PR TITLE
Compass: added MMC3416 driver

### PR DESCRIPTION
**Pull request by tridge**
This driver isn't in use by any of our supported boards yet, but I suspect it will be soon. I'd like it in the tree to prevent bitrot, and to save someone else having to re-implement it. It is a surprisingly tricky driver to get right.
I have an open question with the memsic engineers about the temperature compensation code, so there may be an update once that is clarified

### CODE REVIEW

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        break;
+    }
+
+    case STATE_MEASURE_WAIT2: {
+        uint8_t status;
+        if (!dev->read_registers(REG_STATUS, &status, 1) || !(status & 1)) {
+            break;
+        }
+        uint16_t data1[3];
+        if (!dev->read_registers(REG_XOUT_L, (uint8_t *)&data1[0], 6)) {
```
Comment (muramura) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: i think, change value 6 is sizeof(data1).
value 6 is magic number.

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        return false;
+    }
+
+    dev->set_retries(10);
+    
+    uint8_t whoami;
+    if (!dev->read_registers(REG_PRODUCT_ID, &whoami, 1) ||
+        whoami != 0x06) {
+        // not a MMC3416
+        goto fail;
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: only place with the goto? maybe just release semaphore here and return.

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    const uint16_t sensitivity = 2048; // counts per Gauss, 16 bit mode
+    const float counts_to_milliGauss = 1.0e3f / sensitivity;
+    Vector3f field;
+
+    if (!dev->read_registers(REG_XOUT_L, (uint8_t *)&data, sizeof(data))) {
+        goto check_registers;
+    }
+
+    // this assumes 16 bit output, which gives zero of 32768
+    field(float(data.magx) - zero_offset, float(data.magy) - zero_offset, float(data.magz) - zero_offset);
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: why cast? zero_offset and data.magx are both integer types

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 4]
```diff
@@ -113,41 +111,96 @@ bool AP_Compass_MMC3416::init()
...
+            break;
+        }
+        const uint16_t zero_offset = 32768; // 16 bit mode
+        const uint16_t sensitivity = 2048; // counts per Gauss, 16 bit mode
+        const float counts_to_milliGauss = 1.0e3f / sensitivity;
+        Vector3f field;
+    
+        Vector3f f1(float(data0[0]) - zero_offset,
+                    float(data0[1]) - zero_offset,
+                    float(data0[2]) - zero_offset);
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: changed here but same comment about the cast to float remains.

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 5]
```diff
@@ -102,6 +108,9 @@ bool AP_Compass_MMC3416::init()
     dev->register_periodic_callback(10000,
                                     FUNCTOR_BIND_MEMBER(&AP_Compass_MMC3416::timer, bool));
 
+    // wait 250ms for the compass to make it's initial readings
+    hal.scheduler->delay(250);
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: if you were waiting and doing something.... ok, but just wait, why?  Isn't this just going to slowdown the startup sequence which is serialized?

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 6]
```diff
@@ -168,45 +189,100 @@ bool AP_Compass_MMC3416::timer()
...
+        accum += field;
+        accum_count++;
+        _sem->give();
+    }
+}
+
 void AP_Compass_MMC3416::read()
 {
-    if (!_sem->take_nonblocking()) {
+    if (!_sem->take(0)) {
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: :(

**Comment by lucasdemarchi** [comment]
@tridge added some comments... Github marks some of them  as "outdated", but don't trust them, it's a lie :)

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    
+    /*
+      we use the SET/RESET method to remove bridge offset every
+      measure_count_limit measurements. This involves a fairly complex
+      state machine, but means we are much less sensitive to
+      temperature changes
+     */
+    switch (state) {
+    case STATE_REFILL1:
```
Comment (muramura) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: The description of the switch statement is new. I think that it is better to make it a new description.

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp (hunk 5 above)
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: in testing some callers wanted to see the compass healthy straight after init()

File: libraries/AP_Compass/AP_Compass_MMC3416.cpp [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    }
+    
+    dev->set_device_type(DEVTYPE_MMC3416);
+    set_dev_id(compass_instance, dev->get_bus_id());
+
+    dev->set_retries(1);
+    
+    // call timer() at 100Hz
+    dev->register_periodic_callback(10000,
+                                    FUNCTOR_BIND_MEMBER(&AP_Compass_MMC3416::timer, bool));
```
Comment (OXINARF) [inline libraries/AP_Compass/AP_Compass_MMC3416.cpp]: It returns void now I believe.

**Pull request merged**